### PR TITLE
wip: added good fund manager contract

### PR DESCRIPTION
**Pull request by ordd**
# Description

a contract that collects interest from staking contracts and transfer the interest to the goodReseve. it receives GD interest from goodReserve and transfers it back to the staking contract.

Implements FeeLess scheme

About # (link your issue here)
#100 
#89 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: stakingModel/contracts/GoodFundManager.sol [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        onlyRegistered
+        reserveHasInitialized
+    {
+        // cdai balance of the reserve contract
+        uint256 currentBalance = cDai.balanceOf(address(reserve));
+        // collects the interest from the staking contract and transfer it directly to the reserve contract
+        staking.collectUBIInterest(address(reserve));
+        // finds the actual transferred cdai
+        uint256 actualCDaiGains = cDai.balanceOf(address(reserve)).sub(currentBalance);
+        // mints gd while the interest amount is equal to the transferred amount
```
Comment (sirpy) [inline stakingModel/contracts/GoodFundManager.sol]: check if zero

File: stakingModel/contracts/GoodFundManager.sol [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        (uint256 gdInterest, uint256 gdUBI) = reserve.mintInterestAndUBI(cDai, actualCDaiGains, actualCDaiGains);
```
Comment (sirpy) [inline stakingModel/contracts/GoodFundManager.sol]: check if zeros not need to do some stuff if zero

File: stakingModel/contracts/GoodFundManager.sol [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        // transfers the minted tokens to the given staking contract
+        GoodDollar token = GoodDollar(address(avatar.nativeToken()));
+        token.transfer(address(staking), gdInterest);
+        emit UBITransferred(msg.sender, address(staking), address(reserve), actualCDaiGains, gdInterest);
```
Comment (sirpy) [inline stakingModel/contracts/GoodFundManager.sol]: is this the best event name?
why gdUBI isnt in event?

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/contracts/GoodFundManager.sol (hunk 1 above)
Comment (ordd) [reply stakingModel/contracts/GoodFundManager.sol]: added 41b2a9b38c14d3858a976473ded8d0ca57fd57e8

File: stakingModel/contracts/GoodFundManager.sol (hunk 2 above)
Comment (ordd) [reply stakingModel/contracts/GoodFundManager.sol]: added 41b2a9b38c14d3858a976473ded8d0ca57fd57e8

File: stakingModel/contracts/GoodFundManager.sol (hunk 3 above)
Comment (ordd) [reply stakingModel/contracts/GoodFundManager.sol]: the event name was changed
gdUBI was added
41b2a9b38c14d3858a976473ded8d0ca57fd57e8

**Pull request merged**